atc/creds/pool.go

```diff
@@ -0,0 +1,110 @@
+package creds
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"code.cloudfoundry.org/lager"
+	"code.cloudfoundry.org/lager/lagerctx"
+	"encoding/json"
+)
+
+type VarSourcePool interface {
+	FindOrCreate(lager.Logger, map[string]interface{}, ManagerFactory) (Secrets, error)
+}
+
+type inPoolManager struct {
+	manager        Manager
+	secretsFactory SecretsFactory
+	lastUseTime    time.Time
+}
+
+func (m *inPoolManager) Close(logger lager.Logger) {
+	m.manager.Close(logger)
+}
+
+func (m *inPoolManager) NewSecrets() Secrets {
+	m.lastUseTime = time.Now()
+	return m.secretsFactory.NewSecrets()
+}
+
+type varSourcePool struct {
+	pool map[string]*inPoolManager
+	lock sync.Mutex
+	ttl  time.Duration
+}
+
+func (pool *varSourcePool) FindOrCreate(logger lager.Logger, config map[string]interface{}, factory ManagerFactory) (Secrets, error) {
+	b, err := json.Marshal(config)
+	if err != nil {
+		return nil, err
+	}
+
+	key := string(b)
+
+	pool.lock.Lock()
+	defer pool.lock.Unlock()
+
+	if pool.pool[key] == nil {
+		manager, err := factory.NewInstance(config)
+		if err != nil {
+			return nil, err
+		}
+		err = manager.Init(logger)
+		if err != nil {
+			return nil, err
+		}
+		secretsFactory, err := manager.NewSecretsFactory(logger)
+		if err != nil {
+			return nil, err
+		}
+
+		pool.pool[key] = &inPoolManager{
+			manager:        manager,
+			secretsFactory: secretsFactory,
+			lastUseTime:    time.Now(),
+		}
+	} else {
+		logger.Debug("found-existing-credential-manager")
+	}
+
+	return pool.pool[key].NewSecrets(), nil
+}
+
+func (pool *varSourcePool) Run(ctx context.Context) error {
```

Comment:
I guess renaming the method will bring certain efforts, as the `interface Task` is defined in `atc/lockrunner/runner.go`, which is not specific to GC collector.

From my side, what I can do is to rename `varSourcePool.Run()` to something like `varSourcePool.Clean()`, and add a `varSourcesCollector` under `atc\gc`. But I don't really like that, because it would just introduce some unnecessary duplicate code.

@vito WDYT?